File: frontend/src/types/session.ts
// ─── Session (Mùa giải) ────────────────────────────────────────────────────

export type SessionStatus = 'Draft' | 'Active' | 'Completed' | 'Abandoned';
export type GoalType = 'Bulking' | 'Cutting' | 'Maintaining' | 'Recomp';

export interface Session {
  id: number;
  user_id: number;
  name: string;
  goal_type: GoalType;
  start_date: string;
  end_date: string;
  is_active: boolean;
  status: SessionStatus;
  phases: Phase[];
  created_at?: string;
}

export interface SessionListItem {
  id: number;
  user_id: number;
  name: string;
  goal_type: GoalType;
  start_date: string;
  end_date: string;
  is_active: boolean;
  status: SessionStatus;
  phase_count?: number;
  created_at?: string;
}

export interface SessionCreate {
  name: string;
  goal_type: GoalType;
  start_date: string;
  end_date: string;
  is_active?: boolean;
}

export interface SessionUpdate {
  name?: string;
  goal_type?: GoalType;
  start_date?: string;
  end_date?: string;
}


// ─── Phase (Giai đoạn) ────────────────────────────────────────────────────

export interface Phase {
  id: number;
  session_id: number;
  name: string;
  order: number;
  start_date: string;
  end_date: string;
  target_calories?: number | null;
  target_protein?: number | null;
  target_carbs?: number | null;
  target_fat?: number | null;
  workout_program_id?: number | null;
  workout_program_snapshot?: any;
  description?: string | null;
  created_at?: string;
}

export interface PhaseCreate {
  name: string;
  order?: number;
  start_date: string;
  end_date: string;
  target_calories?: number | null;
  target_protein?: number | null;
  target_carbs?: number | null;
  target_fat?: number | null;
  workout_program_id?: number | null;
  nutrition_score_weight?: number;
  workout_score_weight?: number;
  weight_score_weight?: number;
}

export interface PhaseUpdate {
  name?: string;
  order?: number;
  start_date?: string;
  end_date?: string;
  target_calories?: number | null;
  target_protein?: number | null;
  target_carbs?: number | null;
  target_fat?: number | null;
  workout_program_id?: number | null;
  nutrition_score_weight?: number;
  workout_score_weight?: number;
  weight_score_weight?: number;
}


// ─── DailyLog (Nhật ký hằng ngày) ──────────────────────────────────────────

export interface DailyLog {
  id: number;
  user_id: number;
  phase_id?: number | null;
  log_date: string;
  target_calories_snapshot?: number | null;
  target_protein_snapshot?: number | null;
  target_carbs_snapshot?: number | null;
  target_fat_snapshot?: number | null;
  weight?: number | null;
  calories_in?: number | null;
  protein_in?: number | null;
  carbs_in?: number | null;
  fat_in?: number | null;
  sleep_hours?: number | null;
  work_hours?: number | null;
  fatigue_level?: number | null;
  is_workout_completed: boolean;
  body_images?: string[] | null;
  compliance_score?: number | null;
  compliance_notes?: string | null;
  created_at?: string;
}

export interface DailyLogInlineUpsert {
  log_date: string;
  weight?: number | null;
  calories_in?: number | null;
  protein_in?: number | null;
  carbs_in?: number | null;
  fat_in?: number | null;
  sleep_hours?: number | null;
  work_hours?: number | null;
  fatigue_level?: number | null;
  is_workout_completed?: boolean;
  body_images?: string[] | null;
}

File: backend/app/schemas/session.py
from datetime import date, datetime
from typing import Optional, List
from pydantic import BaseModel, ConfigDict, Field, field_validator


# ─── Session Schemas ─────────────────────────────────────────────────────────

class SessionCreate(BaseModel):
    name: str = Field(min_length=1)
    goal_type: str = Field(default="Maintaining")  # Bulking / Cutting / Maintaining / Recomp
    start_date: date
    end_date: date
    is_active: bool = False
    workout_template_id: Optional[int] = None
    client_id: Optional[int] = None                 # Gán cho học viên (nếu có)


    @field_validator('end_date')
    @classmethod
    def end_after_start(cls, v, info):
        if 'start_date' in info.data and v <= info.data['start_date']:
            raise ValueError('end_date phải sau start_date')
        return v


class SessionUpdate(BaseModel):
    name: Optional[str] = None
    goal_type: Optional[str] = None
    start_date: Optional[date] = None
    end_date: Optional[date] = None


class SessionStatusUpdate(BaseModel):
    status: str  # Draft / Active / Completed / Abandoned


class PhaseRead(BaseModel):
    id: int
    session_id: int
    name: str
    description: Optional[str] = None
    order: int
    start_date: date
    end_date: date
    target_calories: Optional[float] = None
    target_protein: Optional[float] = None
    target_carbs: Optional[float] = None
    target_fat: Optional[float] = None
    workout_program_id: Optional[int] = None
    workout_program_snapshot: Optional[dict] = None

    created_at: Optional[datetime] = None
    model_config = ConfigDict(from_attributes=True)


class SessionRead(BaseModel):
    id: int
    user_id: int
    client_id: Optional[int] = None
    name: str
    goal_type: str
    start_date: date
    end_date: date
    is_active: bool
    status: str
    assigned_by_coach_id: Optional[int] = None
    phases: List[PhaseRead] = []

    created_at: Optional[datetime] = None
    model_config = ConfigDict(from_attributes=True)


class SessionListRead(BaseModel):
    """Lightweight — no nested phases."""
    id: int
    user_id: int
    client_id: Optional[int] = None
    name: str
    goal_type: str
    start_date: date
    end_date: date
    is_active: bool
    status: str
    assigned_by_coach_id: Optional[int] = None
    phase_count: Optional[int] = None

    created_at: Optional[datetime] = None
    model_config = ConfigDict(from_attributes=True)


# ─── Phase Schemas ───────────────────────────────────────────────────────────

class PhaseCreate(BaseModel):
    name: str = Field(min_length=1)
    description: Optional[str] = None
    order: int = 0
    start_date: date
    end_date: date
    target_calories: Optional[float] = None
    target_protein: Optional[float] = None
    target_carbs: Optional[float] = None
    target_fat: Optional[float] = None
    workout_program_id: Optional[int] = None  # ID giáo án gốc để clone

    @field_validator('end_date')
    @classmethod
    def end_after_start(cls, v, info):
        if 'start_date' in info.data and v <= info.data['start_date']:
            raise ValueError('end_date phải sau start_date')
        return v



class PhaseUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    order: Optional[int] = None
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    target_calories: Optional[float] = None
    target_protein: Optional[float] = None
    target_carbs: Optional[float] = None
    target_fat: Optional[float] = None
    workout_program_id: Optional[int] = None


class SuggestNutritionRequest(BaseModel):
    goal: str
    phase_description: str
    gender: Optional[str] = None
    age: Optional[int] = None
    height: Optional[float] = None
    current_weight: Optional[float] = None
    body_fat_percentage: Optional[float] = None
    activity_level: Optional[float] = None

